theano/gof/utils.py

```diff
@@ -217,14 +217,21 @@ def __call__(cls, *args, **kwargs):
         props = cls.__dict__.get('__props__', None)
         if props is None:
             return super(MetaObject, cls).__call__(*args, **kwargs)
+
         key = [cls]
         for arg in args:
             arg = make_hashable(arg)
             key.append(arg)
+            arg_dtype = get_arg_dtype(arg)
+            if arg_dtype:
+                key.extend(arg_dtype)
         for k, v in kwargs.items():
             key.append(k)
             v = make_hashable(v)
             key.append(v)
+            v_dtype = get_arg_dtype(v)
+            if v_dtype:
+                key.extend(v_dtype)
```

Comment:
I'm not sure if it is a wise idea to have `dtype` as a class's key for hashing. But @nouiz told me to remove `name` from the `Elemwise` as it is not supposed to be cached. That was the reason why I started this [PR](https://github.com/Theano/Theano/pull/5071) as a clean up to the `frozendict` PR.